libraries/SPI/SPI.cpp

```diff
@@ -311,53 +319,55 @@ void SPIClass::write32(uint32_t data, bool msb) {
 
 /**
  * Note:
- *  data need to be aligned to 32Bit
- *  or you get an Fatal exception (9)
+ *  data needs to be aligned to 32 bit boundaries like this: char outBuffer[spiBufferSize] __attribute__ ((aligned (4)));
+ *  if not you may get an Fatal exception (9)
  * @param data uint8_t *
  * @param size uint32_t
  */
 void SPIClass::writeBytes(uint8_t * data, uint32_t size) {
+	uint32_t localSize;
     while(size) {
-        if(size > 64) {
-            writeBytes_(data, 64);
-            size -= 64;
-            data += 64;
-        } else {
-            writeBytes_(data, size);
-            size = 0;
-        }
-    }
+        if(size > 64)
+			localSize = 64;
+		else
+			localSize = size;
+			
+        writeBytes_(data, localSize);
+        size -= localSize;
+		if ((size != 0) && (data)) data += localSize;
+	}
 }
 
-void SPIClass::writeBytes_(uint8_t * data, uint8_t size) {
+
+inline void SPIClass::writeBytes_(uint8_t * data, uint32_t size) {
+#ifndef SPI_SKIP_INITIAL_BUSY_CHECK
     while(SPI1CMD & SPIBUSY) {}
-    // Set Bits to transfer
+#endif
+    // Set bits to transfer
     setDataBits(size * 8);
 
-    volatile uint32_t * fifoPtr = &SPI1W0;
-    uint32_t * dataPtr = (uint32_t*) data;
-    uint8_t dataSize = ((size + 3) / 4);
-
-    while(dataSize--) {
-        *fifoPtr = *dataPtr;
-        dataPtr++;
-        fifoPtr++;
-    }
-
+	if(data) {
+		memcpy((uint8_t*)&SPI1W0, data, size);
```

Comment:
OK, a 32-bit alignment check can be done easily during run-time.

But, what exactly shall I do in case the data it is not aligned? The old code has just copied 32-bit of data in a loop which could cause an "unaligment exception".
I could go for byte-wise copying (I fear word-wise as you mentioned solves only half of the problem, doesn't it?), but as memcpy is defined for byte-wise operation by nature as it does not know anything about the data types transferred it should be safe from my point of view even for an unaligned source buffer.

Nearly as stated here: https://github.com/jcmvbkbc/u-boot-xtensa/blob/master/doc/README.unaligned-memory-access.txt

I've tested my implementation on non-aligned memory locations withount any errors by creating an aligned buffer and setting the pointer used for SPI access to the first 4 locations with identical results (see code below):
```c++
char spiIntfOutBuffer[SPI_BUFFER_SIZE] __attribute__ ((aligned (4))); 
uint8_t* spiIntfOutBufferPtr = (uint8_t*)&(spiIntfOutBuffer[3]);
```

So from my point of view there is no need for changes. Or did I miss something which is not covered by my tests?